cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -203,33 +205,41 @@ private Workflow initializeWorkflow() throws Exception {
     if (!Workflow.class.isAssignableFrom(clz)) {
       throw new IllegalStateException(String.format("%s is not Workflow.", clz));
     }
-
     Class<? extends Workflow> workflowClass = (Class<? extends Workflow>) clz;
-    Workflow workflow = new InstantiatorFactory(false).get(TypeToken.of(workflowClass)).create();
-
-    TransactionContext transactionContext = basicWorkflowContext.getDatasetCache().newTransactionContext();
-    transactionContext.start();
-
-    try {
-      if (workflow instanceof ProgramLifecycle) {
-        basicWorkflowToken.setCurrentNode(workflowSpec.getName());
-        ClassLoader oldClassLoader = setContextCombinedClassLoader(workflow);
-        try {
-          basicWorkflowContext.setState(new ProgramState(ProgramStatus.INITIALIZING, null));
-          ((ProgramLifecycle<WorkflowContext>) workflow).initialize(basicWorkflowContext);
-        } finally {
-          ClassLoaders.setContextClassLoader(oldClassLoader);
+    final Workflow workflow = new InstantiatorFactory(false).get(TypeToken.of(workflowClass)).create();
+    if (!(workflow instanceof ProgramLifecycle)) {
+      return workflow;
+    }
+    final TransactionControl txControl =
+      Transactions.getTransactionControl(TransactionControl.IMPLICIT, Workflow.class,
+                                         workflow, "initialize", WorkflowContext.class);
+    if (TransactionControl.EXPLICIT == txControl) {
+      doInitialize(workflow);
+    }
+    basicWorkflowContext.execute(new TxRunnable() {
```

Comment:
Similar to the comment in `MapReduceRuntimeService`, we don't need to run `runtimeStore.updateWorkflowToken` from a transaction.